File: Linked list/Linked list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
/* run this program using the console pauser or add your own getch, system("pause") or input loop */
struct Node{
	char data;
	struct Node* next;
	struct Node* prev;
};

struct Node* start=NULL;
struct Node* cur = NULL;
struct Node* new_node = NULL;
struct Node* tail=NULL;
int i;

void print(struct Node* n){
	while(n!=NULL){
    if(n==cur){
      printf("(%c) ",n->data);}
    else{
		printf("%c ",n->data);}
		n=n->next;
	}
	if(start==NULL){
		printf("리스트가 비었습니다.");
	}
}

void addTail(char ip[]){
  for(i=1;i<strlen(ip);i++){
	if(ip[i]=='+'){
		if(start==NULL){
        new_node=(struct Node*)malloc(sizeof(struct Node));
        new_node->data=ip[i+1];
		new_node->next=NULL;
		new_node->prev=NULL;
		cur=new_node;
        start=new_node;
      }
    	else{
      	for(cur=start;cur->next!=NULL;cur=cur->next);
      	new_node=(struct Node*)malloc(sizeof(struct Node));
      	new_node->data=ip[i+1];
      	cur->next=new_node;
      	new_node->prev=cur;
      	new_node->next=NULL;
      	cur=new_node;
	  }
	}		
	}
	print(start);
}

void add(char ip[]){
  i=0;
			while(i<strlen(ip)){
				if(i==0){
					if(ip[i]=='+'){
						if(start==NULL){
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							new_node->next=NULL;
							new_node->prev=NULL;
							start=new_node;
							cur=new_node;
						}
						else if(cur->prev==NULL){
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->prev=new_node;
							new_node->next=cur;
							new_node->prev=NULL;
							cur=new_node;
							start=new_node;
						}
						else{
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->prev->next=new_node;
							new_node->next=cur;
							new_node->prev=cur->prev;
							cur->prev=new_node;
							cur=new_node;
						}
					}
				}
				if(i>0){
					if(ip[i]=='+'){
						
						if(cur->next==NULL){
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->next=new_node;
							new_node->prev=cur;
							new_node->next=NULL;
							cur=new_node;
						}
						else{
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->next->prev=new_node;
							new_node->prev=cur;
							new_node->next=cur->next;
							cur->next=new_node;
							cur=new_node;
						}
					}
					}
				i+=1;
			}
			print(start);
}

void Nextadd(char ip[]){
  for(i=0;i<strlen(ip);i++){
				if(ip[i]=='+'){
				if(start==NULL){
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							new_node->next=NULL;
							new_node->prev=NULL;
							start=new_node;
							cur=new_node;
						}
				else if(cur->next==NULL){
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->next=new_node;
							new_node->prev=cur;
							new_node->next=NULL;
							cur=new_node;
						}
				else{
							new_node=(struct Node*)malloc(sizeof(struct Node));
							new_node->data=ip[i+1];
							cur->next->prev=new_node;
							new_node->prev=cur;
							new_node->next=cur->next;
							cur->next=new_node;
							cur=new_node;
							}
			}
		}
			print(start);
}

void get_data(char ip[]){
	if(start==NULL){
		printf("리스트가 비어있습니다.");
	}
	else{
	 printf("%c",cur->data);}
  }

void traverse_front_add(char ip[]){
  cur=start;
  start=cur;
  if(start==NULL){
  	printf("리스트가 비어있습니다.");
  }
  else{
  for(i=0;i<strlen(ip);i++){
    if(ip[i]=='N'){
      cur=cur->next;
    }

    if(i==2){
      if(ip[i]=='+'){
        new_node=(struct Node*)malloc(sizeof(struct Node));
        new_node->data=ip[i+1];
        cur->prev=new_node;
        new_node->next=cur;
        new_node->prev=NULL;
        cur=new_node;
        start=new_node;
      }}
    if(i>2){
      if(ip[i]=='+'){
         new_node=(struct Node*)malloc(sizeof(struct Node));
					new_node->data=ip[i+1];
					cur->next->prev=new_node;
					new_node->prev=cur;
					new_node->next=cur->next;
					cur->next=new_node;
					cur=new_node;
      }
    }
	}}
  print(start);
}

void delete(){
if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
  else{
 if(cur->next==NULL&&cur->prev!=NULL){
    cur->prev->next=NULL;
    cur=start;
    start=cur;
  }
  else if(cur->prev==NULL&&cur->next!=NULL){
    cur->next->prev=NULL;
    cur=cur->next;
    start=cur;
    cur=start;
  }
  else if(cur->next==NULL&&cur->prev==NULL){
  	cur=NULL;
    start=NULL;
  }
  else{
    cur->prev->next=cur->next;
    cur->next->prev=cur->prev;
    cur=cur->next;
  }
  print(start);}
}

void traverse_rear_delete(char ip[]){
  for(cur=start;cur->next!=NULL;cur=cur->next);
  for(i=0;i<strlen(ip);i++){
    if(ip[i]=='-'){
      cur->prev->next=NULL;
      cur=start;
    }
  }
  print(start);
}

void replace(char ip[]){
	if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
  else if(cur->prev==NULL){
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[1];
    cur->next->prev=new_node;
    new_node->next=cur->next;
    new_node->prev=NULL;
    cur=new_node;
    start=new_node;
  }
  else if(cur->next==NULL){
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[1];
    cur->prev->next=new_node;
    cur->prev=new_node->prev;
    new_node->next=NULL;
    cur=new_node;
  }
  else{
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[1];
    cur->prev->next=new_node;
    new_node->prev=cur->prev;
    new_node->next=cur->next;
    cur->next->prev=new_node;
    cur=new_node;
  }
  print(start);
}

int data_count(){
	int c=1;
	if(start==NULL){
		printf("0");
	}
	else{
	for(cur=start;cur->next!=NULL;cur=cur->next){
		c+=1;
	}
	printf("%d",c);}
}

void is_member(char ip[]){
  if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
  else{
  int m=1;
  cur=start;
  start=cur;
	while(1){
    if(start->data==ip[1]){
      m=0;
      printf("1");
      break;
    }
    else if(cur->data==ip[1]){
      printf("%d",m);
      break;
    }
    if(cur->next==NULL){
      if(cur->data==ip[1]){
      printf("%d",m);
      break;
      }
      else{
        printf("리스트에 %c 없음",ip[1]);
        break;
      }
    }
    cur=cur->next;
    m+=1;
  }}
}

void clear(){
 	if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
	else{start=NULL;
  	 if(cur->next==NULL&&cur->prev!=NULL){
    cur->prev->next=NULL;
    cur=start;
    start=cur;
  }
  else if(cur->prev==NULL&&cur->next!=NULL){
    cur->next->prev=NULL;
    cur=cur->next;
    start=cur;
    cur=start;
  }
  else if(cur->next==NULL&&cur->prev==NULL){
  	cur=NULL;
    start=NULL;
  }
  else{
    cur->prev->next=cur->next;
    cur->next->prev=cur->prev;
    cur=cur->next;
  }
printf("리스트를 비웠습니다.");} 
}

int is_empty(){
	if(start==NULL){
		return 1;
	}
	else{
		return 0;
	}
}

void trverse_front_get(char ip[]){
  if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
  else{
  int id,y;
  y=1;
  id=atoi(&ip[0])-1;
  cur=start;
  while(cur->next!=NULL){
  	y+=1;
  	cur=cur->next;
  }
  if(id+1>y){
  	printf("인덱스 범위가 초과되었습니다.");
  }
  if(id<0){
  	printf("인덱스가 범위보다 작습니다.");
	    }
  if(id+1<=y&&id>=0){
  cur=start;
  for(i=0;i<id;i++){
    cur=cur->next;
  } 
	printf("%c",cur->data);}}
}

void traverse_front_replace(char ip[]){
  if(is_empty()){
  	printf("리스트가 비어있습니다.");
  }
  else{
  int rid,u;
  u=1;
  rid=atoi(&ip[0])-1;
  cur=start;
  while(cur->next!=NULL){
  	u+=1;
  	cur=cur->next;
  }
  if(rid+1>u){
  	printf("인덱스 범위가 초과되었습니다.");
  }
  if(rid<0){
  	printf("인덱스가 범위보다 작습니다.");
  }
  if(rid+1<=u&&rid>=0){
  cur=start;
  for(i=0;i<rid;i++){
    cur=cur->next;
  }
  if(cur->prev==NULL){
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[2];
    cur->next->prev=new_node;
    new_node->next=cur->next;
    new_node->prev=NULL;
    cur=new_node;
    start=new_node;
  }
  else if(cur->next==NULL){
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[2];
    cur->prev->next=new_node;
    cur->prev=new_node->prev;
    new_node->next=NULL;
    cur=new_node;
  }
  else{
    new_node=(struct Node*)malloc(sizeof(struct Node));
    new_node->data=ip[2];
    cur->prev->next=new_node;
    new_node->prev=cur->prev;
    new_node->next=cur->next;
    cur->next->prev=new_node;
    cur=new_node;
  }
  print(start);}}
}

void upper(){
	printf("%c",cur->data-=32);
}

void lower(){
	printf("%c",cur->data+=32);
}

void Menu(){
	printf("========================menu========================\n");
	printf("addTail                          : t +(data)\n");
	printf("print                            : L\n");
	printf("get data                         : G\n");
	printf("traverse front                   : <(N)\n");
	printf("N은 다음 인덱스로 한 칸 이동을 뜻\n");
	printf("delete                           : -\n");
	printf("add current position             : +(data)\n");
	printf("add next position                : N +(data)\n");
	printf("delete last                      : > -\n");
	printf("add first position               : < +(data)\n");
	printf("replace current position         : =(data)\n");
	printf("get designated position data     : (index)G\n");
	printf("data count                       : #\n");
	printf("is member                        : ?(data)\n");
	printf("replace designated position data : (index)=(data)\n");
	printf("clear the list                   : C\n");
	printf("is empty                         : E\n");
	printf("upper case                       : U\n");
	printf("lower case                       : l\n");
	printf("view menu                        : M\n");
	printf("Quit                             : Q\n");
}
int main() {
	Menu();
	while(1){
		char ip[30]={};
		gets(ip);
		if(ip[0]=='t'){	
	    addTail(ip);
		}
		else if(ip[0]=='+'){
			add(ip);
		}
		else if(ip[0]=='N'){
			Nextadd(ip);
		}
		else if(ip[0]=='L'){
			print(start);
		}
	    else if(ip[0]=='G'){
	      get_data(ip);
	    }
	    else if(ip[0]=='<'){
	      traverse_front_add(ip);
	    }
	    else if(ip[0]=='-'){
	      delete();
	    }
	    else if(ip[0]=='>'){
	      traverse_rear_delete(ip);
	    }
	    else if(ip[0]=='='){
	      replace(ip);
	    }
	    else if(ip[0]=='#'){
	    	data_count();
		}
		else if(ip[0]=='?'){
			is_member(ip);
		}
		else if(ip[0]=='C'){
			clear();
		}
		else if(ip[0]=='E'){
			if(is_empty()){
				printf("TRUE");
			}
			else{
				printf("FALSE");
			}
		}
		else if(ip[0]=='U'){
			upper();
		}
		else if(ip[0]=='l'){
			lower();
		}
		else if(ip[0]=='V'){
			Menu();
		}
		else if(ip[0]=='Q'){
			break;
		}
    else{
      if(ip[1]=='G'){
        trverse_front_get(ip);
      }
      if(ip[1]=='='){
        traverse_front_replace(ip);
      }
    }
	}
}

File: Linked list/Linked list.py
class Node(object):
    def __init__(self, data, prev=None, next=None):
        self.data = data
        self.prev = prev
        self.next = next


class DoubleLinkedList(object):
    def __init__(self):
        self.head = None
        self.cur=None
        self.tail = self.head

    def addTail(self, data):
        if self.head is None:
            self.head = Node(data)
            self.tail = self.head
            self.cur = self.head
        else:
            node=self.tail
            new_node = Node(data,prev=node)
            node.next = new_node
            self.tail = new_node
            self.cur = new_node

    def pt(self):
        if self.head==None:
            print('LinkedList is empty')
        else:
            n=self.head
            while n!=None:
                if n==self.cur:
                    print("(%s) "%n.data,end='')
                else:
                    print("%s "%n.data,end='')
                n=n.next
            print()

    def getdata(self):
        if self.head==None:
            print('LinkedList is empty')
        else:
            print("(%s) "%self.cur.data)
        
    def delete(self):
        if self.head==None:
            print('LinkedList is empty')
            
        else:
            if self.cur.next==None and self.cur.prev!=None:
                self.cur.prev.next=None
                self.tail=self.cur.prev
                self.cur=self.head
                self.head=self.cur
            elif self.cur.prev==None and self.cur.next!=None:
                self.cur.next.prev=None
                self.cur=self.cur.next
                self.head=self.cur
                self.cur=self.head
            elif self.cur.next==None and self.cur.prev==None:
                self.cur=None
                self.head=None
            else:
                self.cur.prev.next=self.cur.next
                self.cur.next.prev=self.cur.prev
                self.cur=self.cur.next
                if self.cur.next==None:
                    self.tail=self.cur
        
    
    def add(self, data):
        if self.head is None:
            self.head = Node(data)
            self.tail = self.head
            self.cur=self.head
        elif self.cur.prev is None:
            new_node=Node(data)
            self.cur.prev=new_node
            new_node.next=self.cur
            new_node.prev=None
            self.cur=new_node
            self.head=new_node
        else:
            new_node=Node(data)
            self.cur.prev.next=new_node
            new_node.next=self.cur
            new_node.prev=self.cur.prev
            self.cur.prev=new_node
            self.cur=new_node
            

    def addd(self, data):
        if self.cur.next is None:
            new_node=Node(data)
            self.cur.next=new_node
            new_node.prev=self.cur
            new_node.next=None
            self.cur=new_node
        else:
            new_node=Node(data)
            self.cur.next.prev=new_node
            new_node.prev=self.cur
            new_node.next=self.cur.next
            self.cur.next=new_node
            self.cur=new_node
            
    def traversefront(self):
        if self.head is None:
            print('LinkedList is empty')
        else:
            self.cur=self.head

    def traversefrontmove(self,n):
        DL.traversefront()
        for i in range(n):
            self.cur=self.cur.next
            if self.cur.next==None:
                self.tail=self.cur
    
    def addfirst(self, data):
        if self.is_empty():
            self.head = Node(data)
            self.tail = Node(None, self.head)
            self.head.next = self.tail
            self.cur=self.head
        else:
            tmp = self.head
            self.head = Node(data, None, self.head)
            tmp.prev = self.head
            self.cur=self.head

    def traverserear(self):
        if DL.is_empty()==1:
            print("LinkedList is empty")
        else:
            self.cur=self.tail

    def replace(self,data):
        if DL.is_empty()==1:
            print("LinkedList is empty")
        else:
            if self.cur.prev==None:
                new_node=Node(data)
                self.cur.next.prev=new_node
                new_node.next=self.cur.next
                new_node.prev=None
                self.cur=new_node
                self.head=new_node
            elif self.cur.next==None:
                new_node=Node(data)
                self.cur.prev.next=new_node
                self.cur.prev=new_node.prev
                new_node.next=None
                self.cur=new_node
                self.tail=new_node
            else:
                new_node=Node(data)
                self.cur.prev.next=new_node
                new_node.prev=self.cur.prev
                new_node.next=self.cur.next
                self.cur.next.prev=new_node
                self.cur=new_node
            DL.pt()
            
    def index(self,data):
        if DL.is_empty()==1:
            print("LinkedList is empty")
        else:
            x=int(data)-1
            y=1;
            self.cur=self.head
            while self.cur.next!=None:
                y+=1
                self.cur=self.cur.next
        
        
            if x+1>y:
                raise IndexError('index exceed')

            if x<0:
                raise IndexError('index is less than range')
            
            if x+1<=y and x>=0:
                self.cur=self.head
                for i in range(x):
                    self.cur=self.cur.next
                    if self.cur.next==None:
                        self.tail=self.cur

    def find(self,data):
        if DL.is_empty()==1:
            print("LinkedList is empty")
        else:
            c=1
            self.cur=self.head
            while True:
                if self.head.data==data:
                    print("%d번째에 %s가 있습니다"%(c,data))
                    c=0
                    break
                elif self.cur.data==data:
                    print("%d번째에 %s가 있습니다"%(c,data))
                    break
                if self.cur.next==None:
                    if self.cur.data==data:
                        print("%d번째에 %s가 있습니다"%(c,data))
                        break
                    else:
                        print("리스트에 %s가 없습니다"%data)
                        break
                self.cur=self.cur.next
                c+=1
        

    def datacount(self):
        if DL.is_empty()==1:
            print("LinkedList is empty")
        else:
            node = self.head
            cnt = 0
            while node is not None:
                cnt += 1
                node = node.next
            print(cnt)

    def clear(self):
        if DL.is_empty()==1:
            print("LinkedList is already empty")
        
        else:
            self.cur=self.head
            if self.cur.next==None and self.cur.prev!=None:
                self.cur.prev.next=None
                self.tail=self.cur.prev
                self.cur=self.head
                self.head=self.cur
            elif self.cur.prev==None and self.cur.next!=None:
                self.cur.next.prev=None
                self.cur=self.cur.next
                self.head=self.cur
                self.cur=self.head
            elif self.cur.next==None and self.cur.prev==None:
                self.cur=None
                self.head=None
            else:
                self.cur.prev.next=self.cur.next
                self.cur.next.prev=self.cur.prev
                self.cur=self.cur.next
                if slef.cur.next.next==None:
                    self.tail=self.cur.next
            self.head=None
            print("list is clear")
            
    def is_empty(self):
        if self.head==None:
            return 1
        else:
            return 0

    def upper(self):
        self.cur.data=self.cur.data.upper()

    def lower(self):
        self.cur.data=self.cur.data.lower()

def menu():
    print("========================menu========================")
    print("addTail                          : t +(data)")
    print("print                            : L\n")
    print("get data                         : G\n")
    print("traverse front                   : <(N)\n")
    print("(N은 다음 인덱스로 한 칸 이동을 뜻)\n")
    print("delete                           : -\n")
    print("add current position             : +(data)\n")
    print("add next position                : N +(data)\n")
    print("delete last                      : > -\n")
    print("add first position               : < +(data)\n")
    print("replace current position         : =(data)\n")
    print("get designated position data     : (index)G\n")
    print("data count                       : #\n")
    print("is member                        : ?(data)\n")
    print("replace designated position data : (index)=(data)\n")
    print("clear the list                   : C\n")
    print("is empty                         : E\n")
    print("upper case                       : U\n")
    print("lower case                       : l\n")
    print("view menu                        : M\n")
    print("Quit                             : Q\n")
        
DL=DoubleLinkedList()
menu()
while True:
    m=input()
    if m[0]=='t':
        for i in range(len(m)):
            if(m[i]=='+'):
                DL.addTail(m[i+1])
        DL.pt()

    elif m[0]=='L':
        DL.pt()
        
    elif m[0]=='G':
        DL.getdata()
        
    elif m[0]=='<':
        n=0
        if len(m)==1:
            DL.traversefront()
        else:
            for i in range(len(m)):
                if i==2:
                    if m[i]=='+':
                        DL.addfirst(m[i+1])
                if i>2:
                    if m[i]=='+':
                        DL.addd(m[i+1])
                if m[i]=='N':
                    n+=1
        if n>0:
            DL.traversefrontmove(n)
        DL.pt()
        
    elif m[0]=='-':
        DL.delete()
        DL.pt()
        
    elif m[0]=='+':
        for i in range(len(m)):
            if i==0:
                if m[i]=='+':
                    DL.add(m[i+1])
            else:
                if m[i]=='+':
                    DL.addd(m[i+1])
        DL.pt()

    elif m[0]=='N':
        for i in range(len(m)):
            if m[i]=='+':
                DL.addd(m[i+1])
        DL.pt()

    elif m[0]=='>':
        if m[2]=='-':
            DL.traverserear()
            DL.delete()
        else:
            DL.traverserear()
        DL.pt()

    elif m[0]=='=':
        DL.replace(m[1])

    elif m[0]=='?':
        DL.find(m[1])

    elif m[0]=='#':
        DL.datacount()

    elif m[0]=='C':
        DL.clear()

    elif m[0]=='E':
        if DL.is_empty()==1:
            print("TRUE")
        else:
            print("FALSE")

    elif m[0]=='U':
        DL.upper()
        DL.pt()
        
    elif m[0]=='l':
        DL.lower()
        DL.pt()

    elif m[0]=='M':
        DL.menu()

    elif m[0]=='Q':
        break
        
    else:
        DL.index(m[0])
        if m[1]=='G':   
            DL.getdata()
        if m[1]=='=':
            DL.replace(m[2])
            
